### DISCUSSION TITLE
Add Rest API / API testing ability to WebdriverIO

### DISCUSSION THREAD

**Discussion by praveendvd** [webdriverio/webdriverio#7187] [Ideas]
**Is your feature request related to a problem? Please describe.**

It would be great if we can integrate API testing into our WebdriverIO platform, but we could do this also by not reinventing the wheel as there are already opensource tools like chakram which looks to be promising : 

 http://dareid.github.io/chakram/jsdoc/index.html
http://dareid.github.io/chakram/

**Describe the solution you'd like**

Allow user to create browser object as chakram or any other tool object and see how we can integrate request and response data into the exciting reporting framework

**Describe alternatives you've considered**

Create a headless chrome driver capability and run chakram in the test framework like mocha, jasmine etc

**Comment by BorisOsipov** [comment]
>It would be great if we can integrate API testing into our WebdriverIO platform

I dont think so. There are enough libs\tools for API testing.

> tools like chakram which looks to be promising

It is abandoned project based on fully deprecated request http client lib.

**Comment by praveendvd** [comment]
@BorisOsipov  True but having to switch between tools for different test levels can be avoided, now we can use webdriverIO for web, electron, mobile, WPF and adding API to the list sounds interesting.

**Comment by christian-bromann** [comment]
The advantage I see of adding API testing capabilities would be the ability to leverage the WebdriverIO ecosystem, its reporters and services. As @praveendvd mentioned the notion of being able to use one tool for multiple types of testing can be appealing. We see it in other frameworks, e.g. [TheIntern](https://theintern.io/), that combine e2e with unit testing capabilities.

Something I believe we should answer is:
- what are current popular API testing tools
- where could WebdriverIO provide an advantage compared to the existing tools
- how should this look like (e.g. if you omit capabilities, you automatically run API tests, or do we have an `$ wdio run e2e wdio.conf.js` vs. `$ wdio run api wdio.conf.js`?

I can imagine some other opportunities here we can leverage API response data to build mock server for e2e tests.

**Comment by ramunask-talech** [comment]
I use axios library (https://github.com/axios/axios) to manipulate test data through API before tests.

**Comment by erwinheitzman** [comment]
@christian-bromann 
- what are current popular API testing tools -> some that I have used: jest + chai-http, Postman, ReadyAPI, Junit + Mockito, Cucumber + WireMock and a colleague of mine is now utilizing Cypress to run API tests which feels very similar to what we are going to achieve here.
- where could WebdriverIO provide an advantage compared to the existing tools -> A single eco-system to work in, we could utilize our new mocking setup https://webdriver.io/docs/api/browser/mock/ and maybe leverage to a new level, those mocks/calls/spies could perhaps be re-used between api/integration/end-to-end tests
- how should this look like (e.g. if you omit capabilities, you automatically run API tests, or do we have an $ wdio run e2e wdio.conf.js vs. $ wdio run api wdio.conf.js? -> I think it might be nice to add a `api` runner which could spawn the most minimalistic headless chrome instance that could be used by all tests

I like this idea and I think that usability would win over people here since yes, you will be losing a little bit of speed/startup time but if it offers you a single eco-system and a nice API to work with, I feel that people would love to use this.

**Comment by praveendvd** [comment]
**what are current popular API testing tools**

- In nodejs space, the testing tool is mostly Postman. There isn't or haven't been many API testing pure libraries that have good acceptance among the tech industry like eg: Rest-assured in Java , or Ressharp in C#. 

- Having a library-based framework can help in defining acceptance criteria in BDD format and encourage multi stake holder collaboration

**where could WebdriverIO provide an advantage compared to the existing tools:**

- Most tools are Either Pure tools with implementation hidden or pure libraries that don't provide a framework.

- WebdriverIO can be a key player in providing a full-fledged Framework that has all boilerplate codes like reporting, API assertions, client setup etc already done for you 

**how should this look like (e.g. if you omit capabilities, you automatically run API tests, or do we have an $ wdio run e2e wdio.conf.js vs. $ wdio run api wdio.conf.js?**

- i think Allow providing empty capabilities {} , if the under lying browser object or type doesn't require one

**Comment by sachin-s-joshi** [comment]
The idea of having API runner similar to wdio runner sounds a plan.

Currently, I am running my APIs within wdio test runner by making my browser headless, I have used axios or node-fetch for calling APIs and parsing response body and data validation.

I tend to use same repo for UI and my APIs for a case where I would call api to make some change and validate the UI after api response. 
Also, if one wants to run only APIs tests it just works fine, only a browser instance is created by default which has no use in pure api test.

**Reply by praveendvd** [reply]
I did the exact same thing

**Comment by Xotabu4** [comment]
UI tests and API tests have very few in common:
+ Test runner can be same (mocha for example)
+ Some parts of API tests can be used as preconditions in UI tests

But instead a lot of things are very different:

- Reporters: no screenshots, traces, but instead much more different attachments might be needed in reports (like request/response payloads for example).
- Project structure: no page objects, no page components in API tests. 
- Configuration: almost noting from wdio.conf.js is applicable to API testing (pre/post hooks like beforeTest/afterTest, bail, retry, specs, exclude options)
- Services: Only few services can be reused for API tests, i did quick check and marked them with green on screenshot
- Assertions: Current wdio-expect is not applicable to API tests
- Multi-remote, repl mode, and many other nice features are not applicable to API tests.

IMHO: API and UI test have very few things in common. Much easier to create separate API testing framework, instead of adapting WDIO to run everything.

![image](https://user-images.githubusercontent.com/3033972/127374473-f59db0db-691a-4805-b691-95b1d50846bd.png)

**Reply by praveendvd** [reply]
@Xotabu4 

**But instead a lot of things are very different:**

- Reporters: no screenshots, traces, but instead much more different attachments might be needed in reports (like request/response payloads for example). 

you can add all these requests, response payload methods etc as just attachments to reports ( its the same as using cucumber and allure report for API testing in other languages like Java/ C#

- Project structure: no page objects, no page components in API tests.

Pages are just S in SOLID principle, its just SIngle responsibility structure which is implemented in API test as separate step definitions for PATH/ API 

- Configuration: almost noting from wdio.conf.js is applicable to API testing (pre/post hooks like beforeTest/afterTest, bail, retry, specs, exclude options)

onprepare , before , after , afterHook everything is applicable for any test level thats why these hooks are provided for unit test frameworks . It depends on what you use it for (its up to the user)

- Services: Only few services can be reused for API tests, i did quick check and marked them with green on screenshot
Assertions: Current wdio-expect is not applicable to API tests

Why would a user want to use services that are not applicable for them ?

Multi-remote, repl mode, and many other nice features are not applicable to API tests.

you can use multi remote as it is !!! one a : capabilities { for browser} , b: { capabilities: {} } for api . which could be used to do api and front end as we do now for other combinations like mobile and browser, etc

IMHO: API and UI test have very few things in common. Much easier to create separate API testing framework, instead of adapting WDIO to run everything.

For API testing we use restassured / restsharp etc with the excisting frameworks and reporting tools that we use for UI. This feature request is to have same ability for webdriverIO

**Reply by praveendvd** [reply]
Also UI tests are just w3c API calls under the hood

**Comment by saurabhkjain** [comment]
Hi, any updates on this? I

**Reply by christian-bromann** [reply]
Are you interested getting involved implementing this?

**Comment by ghost** [comment]
There are multiple ways to add capability to WDIO outside the base solution and the extensibility of WDIO is one of its best features.

For my team we use WDIO as our base solution and we add DB, API etc on top at our choice and timeline. Where a “built in” solution would differ I cannot really say. Maybe through “services” in the config to disable the browser? But for me that wouldn’t work because we use the API calls to change test conditions at runtime or get data like tokens to use in UI testing.

So AXIOS we are using as a dependency for our small API library in our WDIO delivery has delivered what we needed up until now. 

A project has started internally that is pure API and wants to use the solution and still keep within the WDIO run family. This is an issue because the solution we have built is just simplified class and methods to do the work, not a solution like Postman which is built from the ground up to be a complete testing solution or REST which you can build and test APIs.

So I would say the appetite is there for extending the WDIO project to cover API testing more fully but you would need to allow users a base object (like $() for UI) where they could begin their API test patterns.

**Reply by christian-bromann** [reply]
Thanks you for this feedback!

I agree with you, and so does [the community](https://twitter.com/webdriverio/status/1551962131813236738), that there is an interest bringing API testing capabilities to the project. However I have problems figuring out WebdriverIO could bring to the table and why a WebdriverIO driven API test interface is better than [any other dedicated API testing framework](https://dev.to/asaianudeep/api-testing-tools-in-javascript-22d8). Some ideas I was playing with was run these API tests within the browser but these things are also possible with dedicated frameworks.

Long story short: what are the problems an integrated WDIO API solution can solve?

**Reply by ghost** [reply]
For us it’s the “one stop shop”. I would love to show you what we have done with the solution over here. 

For the internal team that want to use WDIO as an API solution, it’s because of the consistency of implementation across the business I am championing. 

We are slowly dragging the teams using Test Cafe, Cypress and Playwright onto the WDIO solution we have built because the needs of the business are greater than those solutions can cater for and WDIO is the resolution for all the pain points. 

As an example, your implementation of DevTools was the big win for us to get Playwright enthusiasts to move over. 

With API there are a lot of different solutions in the market already so the call to recreate something that can run in the WDIO pattern is a difficult one to manage because you wouldn’t want to make an inferior product to what the others offer. 

But as an example your solution could easily implement something similar to Karate and then teams could adopt it safely knowing they have covered their API and UI in the soma me solution set.

**Reply by christian-bromann** [reply]
Thanks for the input @2erix. What do you think about booking an [Open Office Hour](https://webdriver.io/community/openofficehours) slot and meet to talk about some of the requirements and exchange ideas?
